blockchain/types/tx_internal_data.go

```diff
@@ -539,20 +539,37 @@ func NewTxInternalDataWithMap(t TxType, values map[TxValueKeyType]interface{}) (
 	return nil, errUndefinedTxType
 }
 
-func IntrinsicGasPayload(gas uint64, data []byte) (uint64, error) {
+// toWordSize returns the ceiled word size required for init code payment calculation.
+func toWordSize(size uint64) uint64 {
+	if size > math.MaxUint64-31 {
+		return math.MaxUint64/32 + 1
+	}
+
+	return (size + 31) / 32
+}
+
+func IntrinsicGasPayload(gas uint64, data []byte, isContractCreation bool, rules params.Rules) (uint64, error) {
 	// Bump the required gas by the amount of transactional data
 	length := uint64(len(data))
 	if length > 0 {
 		// Make sure we don't exceed uint64 for all data combinations
 		if (math.MaxUint64-gas)/params.TxDataGas < length {
 			return 0, kerrors.ErrOutOfGas
 		}
+		if isContractCreation && rules.IsMantle {
+			lenWords := toWordSize(uint64(len(data)))
```

Comment:
How about using precalculated `length`?
```suggestion
			lenWords := toWordSize(length)
```

Reply:
Thank you! Applied it.